lib/identity_cache/parent_model_expiration.rb

```diff
@@ -4,44 +4,61 @@ module ParentModelExpiration # :nodoc:
-        send(parent_expiration_entry)
+      parents_to_expire = {}
+      add_parents_to_cache_expiry_set(parents_to_expire)
+      parents_to_expire.each_value do |parent|
+        parent.send(:expire_primary_index)
       end
     end
 
-    def expire_parent_cache_on_changes(parent_name, foreign_key, parent_class, only_on_foreign_key_change)
-      new_parent = send(parent_name)
+    def add_parents_to_cache_expiry_set(parents_to_expire)
+      self.class.parent_expiration_entries.each do |association_name, cached_associations|
+        parents_to_expire_on_changes(parents_to_expire, association_name, cached_associations)
+      end
+    end
 
-      if new_parent && new_parent.respond_to?(:expire_primary_index, true)
-        if should_expire_identity_cache_parent?(foreign_key, only_on_foreign_key_change)
-          new_parent.send(:expire_primary_index)
-          new_parent.send(:expire_parent_caches) if new_parent.respond_to?(:expire_parent_caches, true)
-        end
+    def add_record_to_cache_expiry_set(parents_to_expire, record)
+      key = record.primary_cache_index_key
+      unless parents_to_expire[key]
+        parents_to_expire[key] = record
+        record.add_parents_to_cache_expiry_set(parents_to_expire) if record.respond_to?(:add_parents_to_cache_expiry_set, true)
       end
+    end
 
+    def parents_to_expire_on_changes(parents_to_expire, association_name, cached_associations)
+      parent_association = self.class.reflect_on_association(association_name)
+      foreign_key = parent_association.association_foreign_key
+
+      new_parent = send(association_name)
+
+      old_parent = nil
       if transaction_changed_attributes[foreign_key].present?
         begin
-          old_parent = parent_class.find(transaction_changed_attributes[foreign_key])
-          old_parent.send(:expire_primary_index) if old_parent.respond_to?(:expire_primary_index, true)
-          old_parent.send(:expire_parent_caches)  if old_parent.respond_to?(:expire_parent_caches, true)
+          if parent_association.options[:polymorhpic]
+            klass = transaction_changed_attributes[parent_association.association_foreign_key].try(:constantize)
+            klass ||= new_parent.class
+          else
+            klass = parent_association.klass
+          end
+          old_parent = klass.find(transaction_changed_attributes[foreign_key])
         rescue ActiveRecord::RecordNotFound => e
           # suppress errors finding the old parent if its been destroyed since it will have expired itself in that case
         end
       end
 
-      true
+      cached_associations.each do |parent_class, only_on_foreign_key_change|
+        if new_parent && new_parent.is_a?(parent_class) && should_expire_identity_cache_parent?(foreign_key, only_on_foreign_key_change)
+          add_record_to_cache_expiry_set(parents_to_expire, new_parent)
+        end
+
+        if old_parent.try(:is_a?, parent_class)
```

Comment:
`is_a?` is defined on `Kernel` and is available all they down to `BasicObject`. When is this not defined?